content/en/blog/2023/ambient-performance-savings/index.md

```diff
@@ -0,0 +1,131 @@
+
+- [Ztunnels](/blog/2023/rust-based-ztunnel/) are a brand new Istio component written in Rust that are designed to be fast, secure, and lightweight. Ztunnels are deployed per node and serve all pods co-located on the node. They enable the most basic service mesh configurations for L4 features such as mTLS, telemetry, authentication, and L4 authorizations.
+- [Waypoint proxies](/blog/2023/waypoint-proxy-made-simple/) provide L7 mesh features such as VirtualService routing, L7 telemetry, and L7 authorizations policies. Waypoints are still based on Envoy and are deployed per namespace or service account.
+
+These ztunnels and waypoint proxies work in tandem to replace sidecars in the Istio service mesh, and most importantly, unlike sidecars they scale independent to a user's applications pods. So let’s take a closer look at how the two architectures compare in the scenarios above.
+
+## A Closer Look
+
+{{< image width="100%"
+    link="cpu-by-pod.png"
+    caption="CPU usage by pod for sidecar and ambient pods"
+    >}}
+
+{{< image width="100%"
+    link="memory-by-pod.png"
+    caption="Memory usage by pod for sidecar and ambient pods"
+    >}}
+
+{{< image width="100%"
+    link="resource-usage.png"
+    caption="CPU and Memory usage for cluster"
+    >}}
+
+Let’s start by looking at CPU usage by pod. In the sidecar scenarios, the containers utilizing the most CPU resources are the [frontend](https://github.com/GoogleCloudPlatform/microservices-demo/tree/main/src/frontend) service sidecars, which is the main server hit by the load generators and also responsible for making any subsequent calls to other underlying microservices.
+
+In the ambient scenarios, a particular ztunnel's performance depends on the distribution of services across the nodes in the cluster. Ztunnels do use more CPU than a sidecar, but each ztunnel is handling inbound and outbound connections for about 48 application pods whereas a sidecar is responsible for only one.  In the ambient with both ztunnel and waypoint proxy scenario, there is an increase in average CPU usage per ztunnel as traffic is now directed to waypoint proxies. The waypoint proxies themselves consume more resources than their sidecar counterparts.  This is expected since there are only 3 waypoints handling L7 processing for all 48 pods in the namespace.
+
+Next is memory usage by pod. In both ambient scenarios, the L4 ztunnel consumes less memory than a sidecar. In fact, in the L4 only scenario, almost all three ztunnel instances combined consume less memory than a single sidecar. Ztunnels are able to serve multiple applications while consuming fewer resources than sidecars because they are lightweight and written in Rust. Waypoint proxies consume a similar amount of resources as sidecars do, but once again see a minor increase due to the same reasons outlined above for CPU. So what do these per pod usages mean for the totals across the cluster? Everything.
```

Comment:
> because they are lightweight and written in Rust

I think there's a bit more at play here and we should probably help to clarify a bit upfront. The sidecar is actually composed of an istio-agent (written in Go) and Envoy (written in C++). The istio-agent is what establishes the XDS connection to istiod. Envoy then establishes an XDS connection to the istio-agent, which is acting as an XDS proxy to istiod.

So yes, Rust means lightweight and potentially faster than the Go code. Envoy on the other hand is C++, which could potentially be as light and faster than Rust. However, it's likely slower than our ztunnel due to Envoy's architecture not being optimized well for our partiular use case (HBONE). To expand this even further, ztunnel combines the roles of istio-agent and the proxy into a single binary, which eliminates the baggage of multiple processes, multiple XDS connections, and the duplication of memory/CPU that all of that entails.

Even if we keep the discussion of "how sidecars actually work" to a minimum, it would probably go a long way to help explain why we're better across the board.

@howardjohn any thoughts?

Reply:
I agree with everything you said, https://istio.io/latest/blog/2023/rust-based-ztunnel/#a-purpose-built-ztunnel covers this a bit as well. Which really boils down to "purpose built, and designed to have a small footprint".
At a small scale, the agent (which is mostly a fixed cost) is likely a high percentage. In a large cluster where envoy is using like 500mb or something, the agent aspect is probably not particularly relevant, and the other aspects become more important.

Reply:
Would it be sufficient to provide a link to the ztunnel blog given this blog is focusing on resource saving?

Reply:
Yeah that SGTM. Something like "See here for details on why its lightweight" or similar

Reply:
@GregHanson can you address this comment? Thanks!

Reply:
Honestly, I think a short introduction of both sidecar and ztunnel with diagrams for how they differ would go a long way here. Definitely provide the link as well for more information.

Reply:
@nmittler I updated to include a sentence directing users to the blog for some more clarification.  I can also delete the phrase entirely as well.  This section was more for discussing results than comparing the two architectures.  If it is a blocker for you, I can add some images from https://istio.io/v1.15/blog/2022/introducing-ambient-mesh/#istio-and-sidecars to the `Where the Savings Are Coming From` section

Reply:
I really think images and a short overview of how the architectures differ will be essential to the reader understanding the performance difference. I don't think a simple link is enough. We don't really capture this information anywhere today ... at least not with a concise comparison with diagrams.

Reply:
I am very worried about adding architecture differs and explanations here to sidetrack the main purpose of the blog. If we didn't explain the architecture differs in the announcement blog well, we have another opportunity to clarify them in the ambient docs.  @GregHanson showed me the 2 diagrams he built, i think it is too much details to get to pilot-agent, XDS, SDS, etc for this blog.

Reply:
@linsun 

> If we didn't explain the architecture differs in the announcement blog well, we have another opportunity to clarify them in the ambient docs

We don't currently have a good link to give readers for a concise overview. We can't make this blog dependent on some future doc being refactored or re-written just to provide context here. 

>  i think it is too much details to get to pilot-agent, XDS, SDS, etc for this blog.

I strongly disagree. This is a technical blog that requires that the reader understands not only the architecture of ztunnel but how it differs from sidecar. Differences in memory consumption and CPU are in no small part due to not only the language (rust vs golang), but also architecture. The reader needs to understand that to understand the results.

Reply:
I'm not really sure I grok Nathan's specific ask in his comments so I'm not clear if this is agreement/restatement or just another perspective.

I view the salient point as switching from per-pod to per-node (or per-node + per-ns/sa on L7) reduces resource use pretty sharply as stated in the "Where the Savings Are Coming From" section. Language selection and pilot-agent and custom proxy optimized for L4 are arguably all related details but they don't provide the bulk of the savings.

If a diagram is needed I think finding a way to make it simple and to highlight the move from sidecar to ztunnel (optionally +waypoint) for the data plane is what we'd want to focus on in this blog. The rest could be probably be omitted.

I'm not arguing that the text should avoid mentioning these other points but that perhaps we can leave visually depicting them to another blog post.